src/HEVM.hs

```diff
@@ -365,3 +387,112 @@ checkOp (UIntMin _ _) = error "Internal error: invalid in range expression"
 checkOp (UIntMax _ _) = error "Internal error: invalid in range expression"
 checkOp (ITE _ _ _ _) = error "Internal error: invalid in range expression"
 checkOp (IntEnv _ _) = error "Internal error: invalid in range expression"
+
+
+-- | Checks whether all successful EVM behaviors are withing the
+-- interfaces specified by Act
+checkAbi :: SolverGroup -> VeriOpts -> Act -> BS.ByteString -> IO [EquivResult]
+checkAbi solver opts act bytecode = do
+  let txdata = EVM.AbstractBuf "txdata"
+  let selectorProps = assertSelector txdata <$> nubOrd (actSigs act)
+  evmBehvs <- getBranches solver bytecode (txdata, [])
+  let queries =  fmap assertProps $ filter (/= []) $ fmap (checkBehv selectorProps) evmBehvs
+
+  when opts.debug $ forM_ (zip [(1 :: Int)..] queries) $ \(idx, q) -> do
+    TL.writeFile
+      ("abi-query-" <> show idx <> ".smt2")
+      (formatSMT2 q <> "\n\n(check-sat)")
+  
+  results <- fmap (toVRes msg) <$> mapConcurrently (checkSat solver) queries
+  case all isQed results of
+    True -> pure [Qed ()]
+    False -> pure $ filter (/= Qed ()) results
+
+  where
+    actSig (Behaviour _ _ iface _ _ _ _ _) = T.pack $ makeIface iface
+    actSigs (Act _ [(Contract _ behvs)]) = actSig <$> behvs
+    -- TODO multiple contracts
+    actSigs (Act _ _) = error "TODO multiple contracts"
+
+    checkBehv :: [EVM.Prop] -> EVM.Expr EVM.End -> [EVM.Prop]
+    checkBehv assertions (EVM.Success cnstr _ _ _) = assertions <> cnstr
+    checkBehv _ (EVM.Failure _ _ _) = []
+    checkBehv _ (EVM.Partial _ _ _) = []
+    checkBehv _ (EVM.ITE _ _ _) = error "Internal error: HEVM behaviours must be flattened"
+    checkBehv _ (EVM.GVar _) = error "Internal error: unepected GVar"
+
+    msg = "\x1b[1mThe following function selector results in behaviors not covered by the Act spec:\x1b[m"
+
+-- | decompiles the given EVM bytecode into a list of Expr branches
+getBranches :: SolverGroup -> BS.ByteString -> Calldata -> IO [EVM.Expr EVM.End]
```

Comment:
don't we have this in `SymExec.hs` in hevm already?

Reply:
Yes, we do. We can't directly use it because it's local to `checkEquiv` but at some point we should generalise it and use it here as well. Same for some other definitions in this file.